### PR TITLE
Test phases, fixups

### PR DESCRIPTION
**Pull request by shans**
Lots of stuff in this one, sorry. Highlights:

* the test-phases, which I hope we can use for all our experiment testing needs in the future. In particular, capture lets a controlling environment gain access to a log of data.
* adjustments to scheduling based on issues I found while setting up some basic test-phases tests.
* a rough scheduler log that I'll turn into something useful one day.
* adding capture dumping and termination checking to erlnmyr
* adding and promisifying startPipeline and loadPipeline.

### CODE REVIEW

**Comment by shans** [comment]
actually, don't merge this until I've had a chance to review it after #220 lands.

**Comment by shans** [comment]
This is now safe to merge on top of #220

File: core/scheduler.js [hunk 1]
```diff
@@ -112,23 +117,58 @@ function runPhases(phases) {
...
+    waitFor.waitFor(task);
+  taskQueue.push(task);
   startTasks();
+  return task;
 }
 
 function startTasks() {
+  traceScheduler && traceScheduler('----');
+  traceScheduler && traceScheduler('task queue has', waitCount, 'in flight and sees');
+  taskQueue.forEach(function(task) {
```
Comment (dstockwell) [inline core/scheduler.js]: move `traceScheduler &&` to this line

File: lib/test-phases.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  });
+
+function clone(x) {
+  if (typeof x == 'object') {
+    if (x instanceof Array) {
+      var r = [];
+      for (var k = 0; k < x.length; k++) {
+        r[k] = clone(x[k]);
+      }
+      return r;
```
Comment (dstockwell) [inline lib/test-phases.js]: I think this is just `return x.map(clone);`

File: tests/experiment.js [hunk 3]
```diff
@@ -12,19 +12,123 @@
...
+      done();
+    });
+  });
+
+  it('should deal with a harder branch scenario (2)', function(done) {
+    testPipeline(experiment('harder-branch-2'), function() {
+      var events = captureController.getInvocations();
+      assert(events.length == 2);
+      assert(events[0].name == events[1].name, "data was duplicated in branch");
+      if (wentThrough(events[0], 't1To1')) {
```
Comment (dstockwell) [inline tests/experiment.js]: why isn't this deterministic?

**Comment by dstockwell** [comment]
lgtm

File: core/scheduler.js (hunk 1 above)
Comment (shans) [inline core/scheduler.js]: done

File: lib/test-phases.js (hunk 2 above)
Comment (shans) [inline lib/test-phases.js]: yup, think so.

File: tests/experiment.js (hunk 3 above)
Comment (shans) [inline tests/experiment.js]: It is, but I don't know which order is *right* yet. The test can stay flexible until we work that out, I think.

**Comment by shans** [comment]
I signed it!

**Pull request merged**